### PR TITLE
State snapshot improvement

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
- There are some edgecases in which a certain trie node is not saved in the proper storer. The problem is that we will have storers that do not have the complete state.
  
## Proposed Changes
- During the snapshot process, if a trie node is not present in the previous storer (which should have had the entire state), it will also be saved there.

## Testing procedure
- Normal test, but with "*:DEBUG,trie:TRACE" log level, and we will see in the logs that the missing nodes are also saved in the  previous storer.

### CODE REVIEW

File: trie/snapshotTrieStorageManager.go [hunk 1]
```diff
@@ -35,17 +38,40 @@ func (stsm *snapshotTrieStorageManager) Get(key []byte) ([]byte, error) {
 		return nil, errors.ErrContextClosing
 	}
 
-	val, err := stsm.mainSnapshotStorer.GetFromOldEpochsWithoutAddingToCache(key)
+	val, epoch, err := stsm.mainSnapshotStorer.GetFromOldEpochsWithoutAddingToCache(key)
 	if errors.IsClosingError(err) {
 		return nil, err
 	}
 	if len(val) != 0 {
+		stsm.putInPreviousStorerIfAbsent(key, val, epoch)
```
Comment (iulianpascalau) [inline trie/snapshotTrieStorageManager.go:46]: putIfAbsent :D 
brings back so many memories

File: trie/snapshotTrieStorageManager.go [hunk 2]
```diff
@@ -35,17 +38,40 @@ func (stsm *snapshotTrieStorageManager) Get(key []byte) ([]byte, error) {
...
 
 	return stsm.getFromOtherStorers(key)
 }
 
+func (stsm *snapshotTrieStorageManager) putInPreviousStorerIfAbsent(key []byte, val []byte, epoch core.OptionalUint32) {
+	if !epoch.HasValue {
+		return
+	}
+
+	if epoch.Value >= stsm.epoch-1 {
```
Comment (iulianpascalau) [inline trie/snapshotTrieStorageManager.go:58]: so basically, if we are in current epoch 500, epoch will be 499 and we test for epoch 498?
Is this >= correct?

Comment (BeniaminDrasovean) [reply trie/snapshotTrieStorageManager.go:58]: The idea is that if the value was not retrieved from the current storer or previous storer, it needs to be saved to the previous storer.

**Review by iulianpascalau — DISMISSED** [review]

**Review by sstanculeanu — APPROVED** [review]

**Review by sstanculeanu — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**